### PR TITLE
optimized/debugged data pipeline

### PR DESCRIPTION
**Pull request by jackd**
Small adjustments to pointnet data pipeline.

A couple of things that will make a difference to performance under certain configurations:
* fixed bug with point cropping (original cropping was done on the batch dimension)
* removed data augmentation from test set, even when data augmentation is turned on (as per original implementation).

Optimizations:
* moved preprocessing ops into `tf.data.Dataset.map`
* added `prefetch`

I don't imagine these optimizations will provide an observable difference in this case, but I think it is appropriate nonetheless if this is to be used as an example.

### CODE REVIEW

File: tensorflow_graphics/projects/pointnet/train.py [hunk 1]
```diff
@@ -13,14 +13,14 @@
...
 # pylint: disable=missing-function-docstring
-
+import functools
 import tensorflow as tf
 
+from tqdm import tqdm
 from tensorflow_graphics.datasets.modelnet40 import ModelNet40
 from tensorflow_graphics.nn.layer.pointnet import PointNetVanillaClassifier as PointNet
-from tensorflow_graphics.projects.pointnet import augment
+from tensorflow_graphics.projects.pointnet import augment as augment_lib
```
[the remark this answers is not in the document]
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py:22]: What parameter name would you prefer I use for `preprocess`? Coinciding with the `FLAGS` name seemed appropriate, but only possible with this change.

[the remark this answers is not in the document]
Comment (ataiya) [reply tensorflow_graphics/projects/pointnet/train.py:22]: Ah I see what you did (github does not show me the whole file by default like our internal tools, but I just found how to see the full one – without diffs, sadly). 

I would split to these, and then the conflict is gone:
```
FLAGS. augment_rotation 
FLAGS: augment_jitter
```

Also, I guess the preprocess function should be moved in the dataset block (#---- delimiters)?

File: tensorflow_graphics/projects/pointnet/train.py [hunk 2]
```diff
@@ -66,17 +66,18 @@
 # ------------------------------------------------------------------------------
 
 
+def preprocess(example, augment):
```
[the remark this answers is not in the document]
Comment (ataiya) [reply tensorflow_graphics/projects/pointnet/train.py:69]: See previous comment, I did not see this change in context at first.
I see you moved it to a "map" operation.

p.s. Thanks, was planning to do that eventually :)

File: tensorflow_graphics/projects/pointnet/train.py [hunk 3]
```diff
@@ -135,8 +136,15 @@ def evaluate():
   num_examples = info.splits["train"].num_examples
   ds_train = ds_train.shuffle(num_examples, reshuffle_each_iteration=True)
   ds_train = ds_train.repeat(FLAGS.num_epochs)
-  ds_train = ds_train.batch(FLAGS.batch_size)
-  ds_test = ModelNet40.load(split="test").batch(FLAGS.batch_size)
+  ds_train = ds_train.map(functools.partial(preprocess, augment=FLAGS.augment),
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/train.py:139]: can you comment on the role of partial here? not familiar.

File: tensorflow_graphics/projects/pointnet/train.py [hunk 4]
```diff
@@ -135,8 +136,15 @@ def evaluate():
...
+                          tf.data.experimental.AUTOTUNE)
+  ds_train = ds_train.batch(FLAGS.batch_size).prefetch(
+      tf.data.experimental.AUTOTUNE)
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/train.py:142]: #NoNewLines :)

File: tensorflow_graphics/projects/pointnet/train.py [hunk 5]
```diff
@@ -135,8 +136,15 @@ def evaluate():
...
+  ds_test = ModelNet40.load(split="test")
+  ds_test = ds_test.map(functools.partial(preprocess, augment=False),
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/train.py:144]: Here too, if possible

File: tensorflow_graphics/projects/pointnet/train.py (hunk 3 above)
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py:139]: Fixes the values of those parameters in the input call. Equivalent to `lambda example: preprocess(examples, augment=FLAGS.augment)`

File: tensorflow_graphics/projects/pointnet/train.py (hunk 4 above)
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py:142]: what's google's column limit? I'm using `yapf` with `{based_on_style: google, indent_width: 2}`

File: tensorflow_graphics/projects/pointnet/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py:22]: sure. Do we want to rotate during testing? My gut feeling is that's appropriate, but it's not what the original authors did.

File: tensorflow_graphics/projects/pointnet/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py:69]: Regarding performance drop: this perplexes me - how are you profiling? With a GPU? I'm getting a 7.5% performance boost with gtx-1050Ti (7 Hz vs 6.5 Hz). Perhaps if you're running on CPU it's slightly slower?

File: tensorflow_graphics/projects/pointnet/train.py (hunk 5 above)
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py:144]: `lambda example: preprocess(example, augment=False)`

**Comment by julienvalentin** [comment]
Hi @jackd !

It seems that you made a lot of progress on this one; what is the current status?

Best.

**Comment by jackd** [comment]
@julienvalentin sorry, got distracted and never got around to addressing comments. I've just made a small tweak as requested by @ataiya (`augment -> (augment_rotation, augment_jitter))`. I see I can now merge - fairly sure that wasn't an option before. Is that because I accepted contributer invite, or has this been reviewed?

If it's the former and there's still a review process to be had, #312 is a larger PR that encompasses these changes, among other things. The tests there are currently failing just because I'm waiting on #310 . atayia has already gone over much of that. If the overhead to starting a review is high I'd suggest starting there and dropping this one, but if you'd rather smaller PRs then this one's ready to go, and I'll manage the merge conflicts with 312 later.

File: tensorflow_graphics/projects/pointnet/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/train.py]: rename / split done. Implementation does NOT rotate the test data.

**Comment by julienvalentin** [comment]
Hi @jackd 

I don't know who changed rights, but you should not merge on Github right now based on how the Google<->Github pipeline works.

#310 is now in, hopefully you are unblocked!

Best.

File: tensorflow_graphics/projects/pointnet/train.py [hunk 6]
```diff
@@ -130,13 +123,33 @@ def evaluate():
...
+  points = example["points"][:FLAGS.num_points]
+  if augment_rotation:
+    points = augment.rotate(points)
+  if augment_jitter:
+    points = augment.jitter(points)
+  example["points"] = points
+  return example
+
+
 if not FLAGS.dryrun:
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/train.py:137]: need to sync with HEAD, as this flag is now gone? (your change)

File: tensorflow_graphics/projects/pointnet/train.py (hunk 4 above)
Comment (ataiya) [reply tensorflow_graphics/projects/pointnet/train.py:148]: Column limit is 80

File: tensorflow_graphics/projects/pointnet/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ataiya) [reply tensorflow_graphics/projects/pointnet/train.py]: Read my comment above, I didn't see that you were doing this as part of the data loading pipeline, not during training anymore

**Comment by jackd** [comment]
@ataiya if #312 isn't too big to do in one PR I might abandon this one - I've already resolved merge conflicts there, will address your comments there shortly. Happy to do both PRs if necessary, just don't want to waste your time.

**Comment by ataiya** [comment]
yes, you can do a single PR. 
Unfortunately GitHub does not allow dependencies between PRs.
